docs/user/tutorial.rst

```diff
@@ -304,16 +436,286 @@ attributes for reading and setting common headers, but you can always
+would prevent us from doing that. They should actually be written before any unit
+tests or application code, but we wanted to get into Falcon testing before going over
+good TDD practices.
+
+In our case (and in the case of most web applications) the idea behind a functional test
+is running the application as a normal, separate process (e.g. with Gunicorn) and
+then interacting with it as normal clients would - through HTTP calls. Before we
+implement that, it would be useful to add the possibility of configuring the image
+storage directory through environment variables in ``app.py``.
+
+.. code:: python
+
+    def get_app():
+        storage_path = os.environ.get('LOOK_STORAGE', '.')
+        image_saver = ImageSaver(storage_path)
+        return create_app(image_saver)
+
+To run the app with a non-default storage directory, just do:
+
+.. code:: bash
+
+    $ LOOK_STORAGE=/tmp gunicorn 'look.app:get_app()'
+
+Now, put this functional test in a new test file (e.g. ``tests/test_functional.py``):
+
+.. code:: python
+
+    import requests
+
+
+    def test_posted_image_gets_saved():
+        location_prefix = '/images/'
+        fake_image_bytes = b'fake-image-bytes'
+
+        response = requests.post('http://localhost:8000/images',
+                      data=fake_image_bytes,
+                      headers={'content-type': 'image/png'})
+
+        assert response.status_code == 201
+        location = response.headers['location']
+        assert location.startswith(location_prefix)
+        filename = location.replace(location_prefix, '')
+        # assuming that the storage path is "/tmp"
+        with open('/tmp/' + filename, 'rb') as image_file:
+            assert image_file.read() == fake_image_bytes
+
+Running this test isn't ideal. You need to manually start the service beforehand
+(with the proper hardcoded storage path), stop the service and clean up the image
+files afterwards. Of course, you could automate the process of starting, stopping,
+and cleaning after the application. And put that automation into the test code
+itself, hopefully in some fixtures. There's a library that can help with that -
+`mountepy <https://github.com/butla/mountepy>`_.
+
+Anyway, with the new test we could get rid assertions checking the parameters
+with which ``ImageSaver.save`` was called by the POST resource. Well, actually,
+we could get rid of both ``test_post_image`` and ``test_saving_image``, because
+they don't check anything more than ``test_posted_image_gets_saved``. But we can
+do that only because we have a very simple application logic.
+
+Normally, you would check component integration and few more meaningful logic paths
+through the entire application with functional tests, and leave the bulk of testing
+to unit tests. But the actual ratio of unit/functional tests depend entirely on
+applications problem domain and will vary.
+
+After this section we won't be doing TDD anymore as you should have a good grip
```

Comment:
"we won't be doing TDD anymore as you" ==> "we'll omit the TDD instructions, as you..."